pgjdbc/src/test/java/org/postgresql/jdbc/ResourceLockTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2004, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.LockSupport;
+
+class ResourceLockTest {
+
+  @Test
+  void testObtainClose() {
+    final ResourceLock lock = new ResourceLock();
+
+    assertFalse(lock.isLocked());
+    assertFalse(lock.isHeldByCurrentThread());
+
+    try (ResourceLock ignore = lock.obtain()) {
+      assertTrue(lock.isLocked());
+      assertTrue(lock.isHeldByCurrentThread());
+    }
+
+    assertFalse(lock.isLocked());
+    assertFalse(lock.isHeldByCurrentThread());
+  }
+
+  @Test
+  void testObtainWhenMultiThreadedExpectLinearExecution() throws InterruptedException, ExecutionException {
+    CallWithResourceLock callWithResourceLock = new CallWithResourceLock();
+
+    int levelOfConcurrency = 5;
+
+    ExecutorService executorService = Executors.newFixedThreadPool(levelOfConcurrency);
+    try {
+      List<Callable<CounterPair>> callables = new ArrayList<>();
+      for (int i = 0; i < levelOfConcurrency; i++) {
+        callables.add(callWithResourceLock::invoke);
+      }
+
+      // expect linear execution
+      List<Future<CounterPair>> results = executorService.invokeAll(callables);
+
+      Set<Integer> preLockSet = new HashSet<>();
+      Set<Integer> postLockSet = new HashSet<>();
+      for (Future<CounterPair> result : results) {
+        CounterPair entry = result.get();
+        preLockSet.add(entry.preLock);
+        postLockSet.add(entry.postLock);
+      }
+
+      assertEquals(levelOfConcurrency, postLockSet.size()); // linear execution inside resource lock block
+      assertEquals(1, preLockSet.size()); // all threads called invoke before any finish
+
+    } finally {
+      executorService.shutdown();
+    }
+  }
+
+  static final class CallWithResourceLock {
+
+    // wait enough time to allow concurrent threads to block on the lock
+    final long waitTime = TimeUnit.MILLISECONDS.toNanos(20);
+    final ResourceLock lock = new ResourceLock();
+    final AtomicInteger counter = new AtomicInteger();
+
+    /**
+     * Invoke returning 'pre lock' and 'post lock' counter value.
+     */
+    CounterPair invoke() {
```

Comment:
Have you considered using `implements Callable<CounterPair>` or `Supplier<CounterPair>`?
It would alleviate `callables.add(callWithResourceLock::invoke)` to `callables.add(callWithResourceLock)`

Reply:
I changed this to use Callable so I'll mark this as resolved.